build-packages/ts-server-plugin/src/declaration.ts

```diff
@@ -0,0 +1,262 @@
+                }
+
+                return { name, type: CLASS_PLUGIN_PROPERTY_TYPE };
+            }
+
+            if (ts.isMethodDeclaration(target)) {
+                return { name, type: CLASS_PLUGIN_METHOD_TYPE };
+            }
+
+            return undefined;
+        }
+
+        if (
+            ts.isFunctionDeclaration(targetMainType)
+            || ts.isVariableDeclaration(targetMainType)
+        ) {
+            return { name: 'unknown', type: FUNCTION_PLUGIN_TYPE };
+        }
+
+        return undefined;
+    }
+
+    getNamespaceReference(): ts.ReferenceEntry | undefined {
+        return this.ctx.nodeUtils.getReferenceForNode(this.node, this.getTextSpan);
+    }
+
+    getNodesByTargetConfig(config: PartialPluginTargetConfig = {}): ts.Node[] {
+        const { type, name } = config;
+        const targetMainType = this.node.parent.parent;
+
+        if (
+            ts.isFunctionDeclaration(targetMainType)
+            || ts.isVariableDeclaration(targetMainType)
+            || ts.isVariableStatement(targetMainType)
+        ) {
+            if (type && type !== FUNCTION_PLUGIN_TYPE) {
+                return [];
+            }
+
+            return [targetMainType];
+        }
+
+        if (!ts.isClassDeclaration(targetMainType)) {
+            return [];
+        }
+
+        const multipleResults = [];
+
+        for (let i = 0; i < this.heritage.length; i++) {
+            const classDec = this.heritage[i];
+
+            const matchingTargets = this.ctx.nodeUtils.getNodeChildByCondition(classDec, (node) => {
+                if (
+                    !ts.isIdentifier(node)
+                    || (name && node.escapedText !== name)
+                ) {
+                    return false;
+                }
+
+                const { parent } = node;
+                const isStatic = !!parent.modifiers?.some((modifier) => modifier.kind === ts.SyntaxKind.StaticKeyword);
+                const isMethod = ts.isMethodDeclaration(parent) || ts.isPropertyDeclaration(parent);
+                // ^^^ special case for arrow functions !!!
+                const isProperty = ts.isPropertyDeclaration(parent);
+
+                if (!type && (isMethod || isProperty)) {
+                    return true;
+                }
+
+                if (isStatic) {
+                    return type === CLASS_PLUGIN_STATIC_TYPE;
+                }
+
+                if (isProperty) {
+                    return type === CLASS_PLUGIN_PROPERTY_TYPE;
+                }
+
+                if (isMethod) {
+                    return type === CLASS_PLUGIN_METHOD_TYPE;
+                }
+
+                return false;
+            });
+
+            multipleResults.push(...matchingTargets);
+        }
+
+        return multipleResults;
+    }
+
+    getNodeByTargetConfig(config: PluginTargetConfig): ts.Node | undefined {
+        const [result] = this.getNodesByTargetConfig(config);
```

Comment:
I feel like this can fail at some point really easily if `getNodesByTargetConfig` will return empty array